core/src/main/java/com/scalar/db/transaction/consensuscommit/TwoPhaseConsensusCommit.java

```diff
@@ -131,6 +133,19 @@ public void mutate(List<? extends Mutation> mutations) throws CrudException {
 
   @Override
   public void prepare() throws PreparationException {
+    // Fill the read set with records from the write and delete sets if they are unread
+    try {
+      crud.fillReadSetForRecordsFromWriteAndDeleteSetsIfUnread();
+    } catch (CrudConflictException e) {
+      throw new PreparationConflictException(
+          "Conflict occurred while reading unread records in the write and delete sets",
+          e,
+          getId());
+    } catch (CrudException e) {
+      throw new PreparationException(
+          "Failed to read unread records in the write and delete sets", e, getId());
+    }
```

Comment:
Before preparing a transaction, we will fill the read set with records from the write and delete sets if they are unread.